`#![register_{attribute,lint}_tool]` (#3808)
rust-lang/rfcs: text/3808-register-tool.md

[inert]: https://rustc-dev-guide.rust-lang.org/attributes.html#builtininert-attributes

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## For users of external tools

Several official tools let you configure their behavior on specific parts of your code. For example, Clippy lets you use `#[warn(clippy::as_ptr_cast_mut)]` to warn on that lint for a single item, and Rustfmt lets you use `#[rustfmt::skip]` to avoid formatting a single item. You can also do this for external tools that are not provided in the Rust toolchain. See the documentation of those tools for the lints and attributes they support.

To tell `rustc` that you are using an external tool, use `#![register_lint_tool(my_tool)]` or `#![register_attribute_tool(my_tool)]` in your crate root.

If a tool name conflicts with a crate names, you can disambiguate the crate with `::my_tool`:

```rust
#![register_attribute_tool(my_tool)]

extern crate my_tool;

// This is the attribute specified in the crate.
#[::my_tool::attribute]
fn foo() {
    // ...
}

// This is the attribute specified by the tool.
#[my_tool::attribute]
fn bar() {
    // ...
}
```

> Crate-level lints for external tools can be configured in `Cargo.toml`, like any other crate-level lint:

Reviewer: To confirm, this is changing cargo from only allowing specific tools to any?

Do we need to provide a way for people to read this, like `cargo metadata`?

What if tool registration is later added to `Cargo.toml`, should we only limit the `lint` namespace to that? That would likely requre an edition to roll out as it would be a breaking change.

Author: > Do we need to provide a way for people to read this, like cargo metadata?

I think that's up to the cargo team. I am happy to specify it in the RFC but I don't think it's strictly necessary.

> What if tool registration is later added to `Cargo.toml`, should we only limit the `lint` namespace to that? That would likely requre an edition to roll out as it would be a breaking change.

I don't follow. What about that change would be breaking?

Reviewer 2: Right now Cargo prints the following error with this `Cargo.toml`:

```toml
[lints.bevy_lint]
pedantic = "warn"
```

```sh
warning: /Users/bdeep/dev/bevy/bevy_cli/bevy_lint/Cargo.toml: unrecognized lint tool `lints.bevy_lint`, specifying unrecognized tools may break in the future.
supported tools: cargo, clippy, rust, rustdoc
```

The warning hurts the UX of using the `[lints]` table with custom tools, but it doesn't stop the table from being used. The RFC does imply that Cargo should support lint tools without any warnings, which would make adding predefined tools a potentially breaking change. I'm not sure if this is what you're talking about, though.

Author: I misunderstood, you are saying that going from the situation after this RFC (supporting [lints] but not tool registration in Cargo.toml) to the future you are proposing (supporting both, and giving a hard error for lints that didn't register your tool) is a breaking change.

I do not think Cargo should give a hard error if a tool is not registered in Cargo.toml. Partly because that's rustc's job, but partly because there are many different ways to configure crate-level attributes, and we should use the final resolved values of the crate attributes, not just what Cargo knows about. So that avoids this problem altogether.

Author: > imo if the user is using a `lints.<tool>` table without it being registered, it should behave in the exact same way as rustc, whether it errors or something else. Rustc has no visibility into this to do something about it.

i disagree, i think it's exactly the opposite. `lints.tool` tells cargo to pass `-W tool::` flags, at which point rustc has all the info to do something about it. but cargo doesn't have all the info, because it doesn't know the crate level attributes.

> I feel like there is no value in providing support for `lints` in this RFC without a canonical way of accessing the data.

the value is in having a way to configure lints for a workspace without having to duplicate them in the root of each crate.

that said, if you feel strongly about this i will try to come up with a canonical format for `cargo metadata`.

Author: how does this look?
```json
{
  "packages": [/* ... */],
  "lints": {
    "bevy_lint": {
      "lint_name": "allow",
    }
  }
```
for the toml file
```toml
[lints.bevy_lint]
lint_name = "allow"
```

Author: oh, maybe you expect this to show up for each dependency too. we could nest them in that way, it would go under `.packages[*].lints`.

Reviewer: > i disagree, i think it's exactly the opposite. lints.tool tells cargo to pass -W tool:: flags, at which point rustc has all the info to do something about it. but cargo doesn't have all the info, because it doesn't know the crate level attributes.

Why would we convert all tools to rustc flags?  The way the RFC is written, it sounds like the tools would not be running inside of rustc.

> the value is in having a way to configure lints for a workspace without having to duplicate them in the root of each crate.

My concern more is how are tools expected to access the lint levels from `Cargo.toml`?

> how does this look?

lints exist at the
- workspace level (well, they will when we add cargo lints)
- package level

> oh, maybe you expect this to show up for each dependency too. 

We have this weird tension in `cargo metadata` about giving people the information they need without slowing things down.  Lints are going to add a lot to the output.  I wonder if we can skip dependencies but what if a tool does need it?

Reviewer 2: I don't think adding `cargo metadata` support to the lints table should be a concern for this RFC. I don't believe it is useful enough for the amount of bikeshedding it requires.

Tools are free to use `cargo locate-project [--workspace]` to find the crate and workspace manifests, then deserialize the `[lints]` table with `serde` + `toml`. This solution is easy to implement and benefits the projects by letting them read custom lint configuration:

```toml
[lints.my_tool]
my_lint = { level = "warn", foo = 3, bar = false }
```

Not all tools even need this. `rustc` drivers already have the proper flags passed to them as `--warn my_tool::my_lint`, so most of the time they won't even need direct access to `[lints]`.

Adding `cargo metadata` support is a nice-to-have, but by no means should it block this RFC.

Author: i've figured out the confusion here. @epage yes, you're right - passing lint flags to rustc does nothing. we need to pass the lint flags to the *tools* instead. but the external tools aren't controlled by cargo (that's the whole point) and so cargo doesn't know how to pass the flags.

i think the right approach here is
1. strip out all the cargo stuff from this RFC, except maybe tool registration
2. change the "authors of tools" guide to suggest using `[metadata.your_tool.lints]` to configure lints crate-level tools

does that sound reasonable to you?

Reviewer: > change the "authors of tools" guide to suggest using [metadata.your_tool.lints] to configure crate-level lints for tools

I see this as a short term solution.  If we're going with that kind of short-term solution there, should we do the same for registration?

Reviewer 3: I agree i think it makes more sense to strip out all the cargo stuff for this RFC and do it in a follow up even if registering tools would be a big QoL improvement.

Author: i have moved all the cargo features to "future possibilities". i don't think there is an urgency to land them now.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

The tool prelude is separated into the tool attribute prelude (which is in the type namespace) and the lint prelude (which is only active inside lint controls).

The tool attribute prelude is changed to take precedence over the [Extern prelude]. The tool attribute prelude is only considered when resolving in an attribute macro; types in other positions, such as `fn foo() -> rustfmt::x`, do not consider the tool prelude. This is to avoid unnecessary breakage when the meaning is clear.

In the context of an attribute, ambiguity between a tool name and a local item is always a hard error. For example, this code would error:

```rust
#![register_tool(name)]

mod inner {
    mod name {
        // Import the derive macro.
        use Clone as x;
    }

    #[name::x]        // ERROR: ambiguous
    #[name::y]        // ERROR: ambiguous (even though y is not present)
    #[self::name::x] // OK
    fn f() {}
}
```
This is in order to not require external tools to perform name resolution. This restriction may be relaxed in the future to favor tool names.

> The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

Author: I don't see why not. `--crate-attr` is effectively a built-in macro, and I explicitly want to allow that. Using other macros seems fine as long as they expand to a crate-level attribute.

> The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

Reviewer 4: I think that should already be covered by usual restricted shadowing rules.

> The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

Author: > #![foo::bar]

to be clear: are you asking about a proc-macro defined in a crate called `foo`? that's disallowed today, proc-macros can't be crate-level attributes.

if you mean a tool named `foo` then that's an inert attribute, not a macro.

> The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

Reviewer 4: (In any case crate-level macros are still extremely unstable.)

> `#[no_implicit_prelude]` keeps its current behavior, i.e. it removes all tools that are in the prelude by default. However, those tools can now be explicitly added back with e.g. `#![register_attribute_tool(rustfmt)]`.

Author: I will specify that `no_implicit_prelude` does not affect tools.

Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).

Modules in the first path of an attribute (e.g. `#[unregistered::name]`) are assumed to be a crate if they cannot be resolved, and therefore give a hard error if not registered.

Registering a tool multiple times is an error. This makes registering a predefined tool (`clippy`, `miri`) using `#![register_*_tool(...)]` an error unless `#![no_implicit_prelude]` is specified.

[tool attributes]: https://doc.rust-lang.org/nightly/reference/attributes.html#tool-attributes
[`unknown_lints`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unknown-lints

# Drawbacks
[drawbacks]: #drawbacks

This makes the rules for name resolution even more complicated.

This runs the risk that external tools will add attributes that change the meaning of the code, such that the behavior is different when the tool is present. There is not much we can do about this other than to ask tool authors not to do that.

> This introduces a new "meta-breaking" concern: once this is stabilized, adding a new tool namespace, like we did for `diagnostic` and `rust_analyzer`, becomes a breaking change. I think in practice there will be few enough external tools that this will not be a major concern and we can just pick a name that isn't already in use. Note that, because `![register_*_tool(...)]` for a built-in tool errors, the user will see immediately what went wrong, instead of getting confusing errors from the attributes in the tool namespace.

Reviewer: Should we pre-emptively reserve some namespaces?

Author: That seems like a good idea - at least `metadata` is something I'd like to reserve. It's much easier to reserve them now than to try and add them later.

Reviewer 4: I'd rather go the opposite way and remove all the hardcoded tools on the next edition, and always require registering them explicitly.
There are zero reasons for rustc to know about, e.g. `rust_analyzer`.

Reviewer 4: >at least `metadata` is something I'd like to reserve

This reminded me of something - https://github.com/rust-lang/rust/pull/75802.

Author: oh, "adding a new namespace has to be over an edition boundary" seems very reasonable, yes. that avoids having to treat built-in tools differently from registered tools.

[lang concern]: https://github.com/rust-lang/rust/issues/66079#issuecomment-1010266282

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- We could "just not do this". That makes it harder to write external tools, and in practice just means that people use `cfg_attr` instead of a namespace, which seems strictly worse.
- We could relax the constraint that tool names cannot conflict with local items. This requires tools to do name resolution; but in practice I do not think we can expect tools to do this, and we must assume that the tool will behave differently than the compiler (`rustfmt` already does this today).
- We could add a syntax to disambiguate tool names from local items. That would add inconsistency with the existing built-in tool attributes, and requires tool authors to parse both the new and existing syntax.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Author: to be clear, you are suggesting that the following be a hard error?
```rust
#![register_attribute_tool(pyo3)]
extern crate pyo3;
#[pyo3::pymodule]  // error: ambiguous between tool and crate
struct Foo {}
```
I think that would not be ideal, but I guess I could live with it as long as `#[::pyo3::pymodule]` is still allowed. I feel somewhat strongly that `#[::pyo3::pymodule]` should not give an error, the `extern crate as` workaround is not very well known. even if people use it as `my_pyo3 = { package = "pyo3" }` in Cargo.toml it's still a bit indirect.

cc @mejrs

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Reviewer 4: >to be clear, you are suggesting that the following be a hard error?

Yes, just because if it's an error initially, then in the future we'll be able to backward compatibly do any of the resolution changes from this RFC (precedence, subnamespaces) if really necessary.

`#[::pyo3::pymodule]` should certainly work, there's no ambiguity there, `::foo` only looks into extern prelude.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Author: I thought of another complication. Say we want to add more built-in tools in the future, or to reserve a tool namespace today. If adding a built-in tool causes breakage because it makes crate names ambiguous, suddenly it's much harder to extend the namespace.

I guess we could have different rules for built-in tools than for registered tools, but that kinda sucks.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Reviewer 5: > I think that would not be ideal, but I guess I could live with it as long as `#[::pyo3::pymodule]` is still allowed.

I've thought about this and and having ` #[::pyo3::pymodule]` work in this case but not `#[pyo3::pymodule]` seems like a bad idea to me. I think it'd be better if they *both* errored or if both worked. If we allow the former then it's gonna be one of those gotchas that *everyone* runs into. "Oh yeah you have to prefix the macro path with `::` because otherwise it's ambiguous between the tool and the macro". I think I'd like to have them named the same but it might be better if authors are forced to choose different names.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Reviewer 5: > you as the author can choose a different name for your tool

I do not disagree in principle, but I'm just a bit paranoid of the situation where someone chooses the crate name as their tool name without thinking about it very much and it ends up being awkward later on. Imagine a situation like:
- author creates a library (without proc macros)
- author creates a tool attribute with the same name, this works fine because of no proc macros
- author later decides to add proc macros
- (without prefix `::`) these macros collide with the tool name
- the author would have to decide to either rename the tool or tell the users to prefix the macros. The latter sounds like a noob trap.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Author: I don't think we can prevent this situation. The last thing still happens even if we prevent disambiguating the tool, we just *force* you to rename, we are strictly limiting your options.

Consider a different case: someone writes a crate, someone else writes a tool, they don't know about each other. you combine them and they conflict. I don't think we should force you in this case to use `extern crate overlap as my_overlap`, that seems strictly worse than allowing you to disambiguate at the call site.

> - We could drop the change to name resolution, such that extern crates continue to take precedence over tools. This allows no way to use a tool and a crate with the same name, unless we add a new syntax.

Reviewer 6: I think the RFC was updated to "always error" so @petrochenkov's original concern is resolved for now. Is that right?

- Projects may wish to have both a proc-macro crate with lints and a CLI with lints. To allow this, we would require `proc_macro_lint` to create an exhaustive list of lints that can be created, such that we can still run `unknown_lints` and do not need to create a new cooperation mechanism between `proc_macro_lint` and `register_lint_tool`, nor to require users of the project to distinguish the two with `::project` (see immediately above). We might still run into difficulty if the proc-macro lint namespace is only active while the proc-macro is expanding; it depends on how `proc_macro_lint` is specified. But I think it's ok to delay that discussion until `proc_macro_lint` gets an RFC.
- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
- Once [expression attributes] are stabilized, this would also allow tool attributes on expressions.

> - We could support registering tools through `Cargo.toml`, which Cargo would then pass to `rustc` using `--crate-attr` (similar to `--check-cfg`). We would have to consider how this interacts with duplicate registration; we don't want to error if a tool is configured both through Cargo.toml and in source code.

Reviewer: Covering every crate in a package can be annoying otherwise which may be worth calling out.

Author: I think I'm going to remove the hard error on duplicate registration and only error if people try to register a built-in tool, which resolves this nicely. I'll also specify a mechanism for Cargo to pass this to rustc. how does `[package] tools = ["c2rust", "bevy_lint"]` sound?

Reviewer: Creating a new top-level table is going to need a lot more elaboration, especially if its as generic as `tools`.

Should this RFC be split into two parts?
- rustc side
- cargo side

#[some_tool::attribute]
fn bar() {
    // ...
}
```

Crate-level lints for external tools can use `#![warn(some_tool::lint_name)]`, like any lint.
Tools may also support a custom configuration format that allows you to control lints for your whole workspace at once.
Consult the documentation of the tool you use.

[Kani]: https://github.com/model-checking/kani
[`bevy_lint`]: https://thebevyflock.github.io/bevy_cli/bevy_lint/
[Bevy game engine]: https://bevyengine.org/

## For authors of external tools

The Rust language can be extended and analyzed using external tools. If your tool can parse Rust, you may wish to allow configuring it at sub-crate levels (e.g. individual functions, types, and modules). To reuse the same syntax as the official tools, like Clippy and Rustfmt, instruct your users to add `#![register_lint_tool(your_tool)]` (if your tool only adds new lints) or `#![register_attribute_tool(your_tool)]` (if your tool only adds new attributes). If your tool supports both lints and attributes, use `#![register_tool(your_tool]`. Then, instruct your users to add either `#[warn(your_tool::your_lint)]` or `#[your_tool::your_attribute(your_tokens)]` as appropriate.

> We do not specify a syntax for crate-level configuration. We suggest using `[package.metadata.your_tool]` in Cargo.toml.

Reviewer: ```suggestion
We do not specify a syntax for package-level configuration. We suggest using `[package.metadata.your_tool]` in Cargo.toml for now.  See [Future Possibilities]
```
(technically the attribute is crate-level and yes our terminology is confusing)

fn foo ( ) { }
```

### Semantics

The tool prelude is separated into the tool attribute prelude (which is in the type namespace) and the lint prelude (which is only active inside lint controls).

The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

`register_attribute_tool` and `register_lint_tool` are idempotent; duplicating the attribute has no effect.

Crate-level macros such as `#![cfg_attr(...)]` may expand to `register_*_tool`, subject to all existing rules for macro expansion.

Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].

Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.

The `rustc` tool namespace is currently reserved and will continue to be reserved after this RFC, i.e, `register_*_tool(rustc)` is an error.

> Additionally, any tool namespace starting with `rustc`, such as `rustc_lint`, is reserved.

Reviewer 7: does this include names like `rustcat`

Author: yeah. i am not a fan of the current attribute gating but i don't want to touch it in this rfc.

Ambiguity between a tool name and any other name in the type namespace is always a hard error. For example, this code would error:

```rust
#![register_tool(name)]

extern crate name;
#[name::skip] // ERROR: ambiguous
#[::name::skip] // OK

> fn foo() {}

Author: > I'm not understanding 100% the reasoning behind denying the ambiguous inert tool attribute vs. a genuine attribute
> A tool (inert) attribute is just that, an inert / nothing-doing syntactical marker that the compiler is to let slide. But a genuine attribute, is just as much a syntactical marker as an inert attribute; so I do wonder about non-compiler-name-res-using tools such as rustfmt not making a difference between the two.

> But I imagine that if a tool were to hook itself onto compiler machinery so as to actually involve name res and semantics, then a difference between a genuine attribute and an inert one could be made.

> In which case, denying the ambiguity makes sense.

@danielhenrymantilla i think you are imagining that the *tool* is responsible for checking there is no ambiguity. that is not the case. the *compiler* is responsible for checking there is no ambiguity, precisely because we want to allow tools that don't do name-res to still use attributes. it's ok if tools accept more code than the compiler, but i think it would be quite bad if the tool interpreted the code differently than the compiler (this is the same reason i have a big "please don't use attributes that change the meaning of the code" mention in the "tool author guide" section).

> But all this may hint at some mention of the distinction between semantics-capable tools (e.g., able to resolve cfgs), vs. purely-syntactical ones being warranted? 🙂

i think distinguishing here is unnecessary complexity, and if we can get away with it we should just use the same rules for all of them. the exception here is between attribute_tool and lint_tool: since they don't affect name res for each other, we can make it easier to add lint-only tools that have the same name as an existing crate without causing breakage. but i don't think we should add more types of tools than necessary.

    #[self::name::x] // OK
    fn f() {}
}
```
This is in order to not require external tools to perform name resolution. This restriction may be relaxed in the future to favor tool names.

To disambiguate a local item, developers may use `self::` to force resolution in the current module.
To disambiguate a crate, developers may write `#[::rustfmt::skip]`, which forces resolution in the [Extern prelude].
To disambiguate an item in the standard prelude (e.g. `Clone`), developers may use `std::prelude::v1::Clone`.
There is no way to disambiguate the tool; developers must use `extern crate rustfmt as my_rustfmt` (or the equivalent in Cargo.toml) to avoid the ambiguity.
As a quality of implementation issue, errors should mention one of the above workarounds.

[inert]: https://rustc-dev-guide.rust-lang.org/attributes.html#builtininert-attributes

Modules in the first path of an attribute (e.g. `#[unregistered::name]`) are assumed to be a crate if they cannot be resolved, and therefore give a hard error if not registered.

Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).

> `#![no_implicit_prelude]` does not affect tools.

Reviewer 7: What does this sentence mean? 

(1) If by "tool" you mean "tool name", today `#![no_implicit_prelude]` does skip the tool prelude and thus remove all the tool names from the attributes (*but not lints*, perhaps a bug), so the "tool names" do get affected.
```rust
#![no_implicit_prelude]
#[deny(clippy::identity_op)] // unexpected, won't raise E0710 (unknown tool name)
#[rustfmt::skip] // expected, error[E0433]: failed to resolve: use of unresolved module or unlinked crate `rustfmt`
const X: u8 = 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 +0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0 + 0;
```

(2) If by "tool" you mean "tool invocation" then yes, the tool does not need to respect `#![no_implicit_prelude]`. For instance `rustfmt` recognizes the `#[rustfmt::skip]` here just fine. But tools that rely on `rustc` such as `clippy` will still be affected by E0433.

(3) Or "tool" here just mean the `register_tool` call and that the following behaves the same with or without `#![no_implicit_prelude]`?

```rust
#![no_implicit_prelude]
#![register_tool(bevy)] // always ok
#![register_tool(rustc)] // always error
```

Author: I do mean 1), tool name. The RFC is specified in terms of the behavior that will be stabilized, not in terms of the current implementation.

[`clang-tidy`]: https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics
[`pylint`]: https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html#block-disables
[`eslint`]: https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1
[`review`]: https://github.com/Bogdanp/racket-review#usage
[Roslyn analyzers]: https://johnnyreilly.com/eslint-your-csharp-in-vs-code-with-roslyn-analyzers#deactivate-linting-partially
[gcc]: https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html
[vendor attributes]:  https://en.cppreference.com/w/cpp/language/attributes
[c-sharp attrs]: https://learn.microsoft.com/en-us/dotnet/standard/attributes/applying-attributes
[active]: https://rustc-dev-guide.rust-lang.org/attributes.html#non-builtinactive-attributes
[`Resyntax`]: https://docs.racket-lang.org/resyntax/Refactoring_Rules_and_Suites.html#(part._.Exercising_.Fine_.Control_.Over_.Comments)

# Unresolved questions
[unresolved-questions]: #unresolved-questions

How does this interact with [proc-macro lints][`proc_macro_lint`]?

# Future possibilities
[future-possibilities]: #future-possibilities

> - We could allow registering tools in Cargo.toml (with a `package.tools` field?). This would avoid duplicating tool registration for each crate in the workspace. This depends on [`--crate-attr`] being stabilized.

Reviewer: We should also consider this process allowing registering tool-specific `cfg`s (e.g. `cfg(kani)`)

We might want to consider a way for tools to define these entries and for users to pull them in somehow.

A tools table may have enough overlap with [runnable dependencies](https://github.com/rust-lang/cargo/issues/2267) that we may want to develop one with the other in mind.

We need to consider whether this table is a `workspace` table like `[profiles]` or a package table.  For the former, we should auto-copy it to the package on publish like we do `workspace.resolver`.  If the latter, we should consider whether workspace inheritance should be allowed.

Author: > We should also consider this process allowing registering tool-specific cfgs (e.g. cfg(kani))

This seems reasonable, but not all tools support cfgs, so doing it for all tools defeats the point of check-cfg. So we would need some way to configure the behavior.

I would be extremely surprised to see someone building an external tool as a dependency. I can mention it I guess but I don't know that it makes sense to go out of our way to support it.

I don't have strong opinions on the other things, but I will mention them in future possibilities.

> - We could make `[lints]` support external tools as a first-class feature. This needs a way for tools to read the metadata out of `Cargo.toml` (e.g. `cargo metadata`, or just parsing the toml file), because cargo does not drive external tools. Additionally, we cannot guarantee that tools will actually read the metadata.

Reviewer: We could have a `[lints]` entry automatically register the tool, even if the table is empty.

- We could allow attribute macros to register a [derive helper], so that they can emit other attributes for an external tool.
- We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.
- Once [expression attributes] are stabilized, this would also allow tool attributes on expressions.
- Some existing attributes, such as [`coverage`], have exactly the semantics of a tool attribute: they add additional meaning when a specific feature or flag is enabled, and ignored otherwise. They could use this mechanism (over an edition boundary, as described above).

> 

Reviewer 6: ```suggestion
- We can allow defining tool attributes in crates as inert declarative macros, which would enable IDE tooling to provide suggestions and syntax-level checking of the attributes. Taking advantage of this would require users to move away from register_tool and use a crate dependency with the name of the tool instead.
  - One danger of accepting this proposal now is that tools will use the crate name for their binary instead of saving it for an eventual support library that can declare attribute macros and potentially other things. This could be mitigated with explicit guidance to use another crate name. For example, kani uses [kani-verifier](https://crates.io/crates/kani-verifier) for the binary while reserving [kani](https://crates.io/crates/kani) for an eventual library.
  - This is somewhat akin to the way the [proc_macro_lint](https://github.com/rust-lang/rust/pull/135432) proposal puts declared lints in the macro namespace.
- We can allow tool attributes to be declared using Rust types. This would the use of versioned, structured metadata that can be shared among different tools and reflection APIs.
  - This can be combined with tool attribute macros to present more interesting syntax options.

```

- Feature Name: `register_tool`
- Start Date: 2025-03-22
- RFC PR: [#3803](https://github.com/rust-lang/rfcs/pull/3808)
- Rust Issue: [rust-lang/rust#66079](https://github.com/rust-lang/rust/issues/66079)

# Summary
[summary]: #summary

> This RFC adds two new attributes:

Author: fixed, thanks

Note that currently, loading a crate completely prevents using a tool attribute with that name. Consider this program:

```rust
extern crate rustfmt; // or --extern rustfmt

#[rustfmt::skip]  // ERROR could not find `skip` in `rustfmt`
fn foo ( ) { }
```

### Semantics

The tool prelude is separated into the tool attribute prelude (which is in the type namespace) and the lint prelude (which is only active inside lint controls).

The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.

`register_attribute_tool` and `register_lint_tool` are idempotent; duplicating the attribute has no effect.

Crate-level macros such as `#![cfg_attr(...)]` may expand to `register_*_tool`, subject to all existing rules for macro expansion.

Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].

> Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.

Reviewer 8: Where will this list be maintained?

Also, today you can't add any attribute that starts with `rustc`. Will that still be the case?

Reviewer 8: From what I understand, today you cannot add a tool attribute that includes the prefix `rustc`, for example, you can register a tool `rustc_foo`, but the compiler will reject the attribute `#[rustc_foo::bar]`.

Author: oh, i see. this is actually not coming from the register_tool code, this is the same code that rejects `#[rustc_baz]`.

i think it is fairly reasonable to keep this reserved, but i can make sure it's rejected earlier in `register_tool` so that people aren't confused.

Reviewer 4: >but i can make sure it's rejected earlier in register_tool so that people aren't confused.

I don't think it makes sense to prevent this, there are many other ways to make an unusable attribute that are not rejected, like `mod rustc_something { pub use my_attribute; }` or `use my_attribute as rustc_something;`, but we do not prohibit modules or imports containing `rustc` preventively.

Reviewer 8: As long as we are consistent. Either it works without any hacks or it fails early on.

- Feature Name: `register_tool`
- Start Date: 2025-03-22
- RFC PR: [#3803](https://github.com/rust-lang/rfcs/pull/3808)
- Rust Issue: [rust-lang/rust#66079](https://github.com/rust-lang/rust/issues/66079)

# Summary
[summary]: #summary

This RFC adds two new attributes:
- `#![register_lint_tool(tool_name)]` allows controlling namespaced lints with `#[warn(tool_name::lint_name)]`

> - `#![register_attribute_tool(tool_name)]` allows using tool names in [inert attributes][inert] with `#[tool_name::attribute_name(token_tree)]`.

Reviewer 8: May I suggest we use `#![register_tool_{attribute|lint}` instead? I was also wondering if we could have a `#[register_tool]` which would add the tool to both namespaces, the attribute and lint.

Reviewer 2: I worry that `register_tool_{attribute|lint}` will imply we are registering _individual_ attributes or _individual_ lints, when that is not the case. If someone sees the attribute `#![register_tool_lint(...)]`, I worry they'll do this:

```rust
#![register_tool_lint(bevy::panicking_methods)]
#![register_tool_lint(bevy::missing_reflect)]
#![register_tool_lint(bevy::duplicate_bevy_dependencies)]
// ...
```

I think the name `register_lint_tool` much better signals that it's for registering a _tool_, not a _lint_ or _attribute_.

Reviewer 8: Gotcha... That makes sense. Thanks

Reviewer 7: unless we allow nested tool name like `com::example::tool`[^1], I don't think `#![register_tool_lint(bevy::panicking_methods)]` will compile because the tool name can only be an `$:ident` not a `$:path`, so it is actually not an issue.

[^1]: `#![warn(com::example::tool::lint)]` is interpreted as having a tool name `com`.

Author: sure, but why not avoid the confusion in the first place?

Reviewer 7: > sure, but why not avoid the confusion in the first place?

IMO `#![register_tool(clippy)]` is also confusing, it sounds like it will make `rustc` run the "tool" `clippy` by "registering" it to the compilation process. The standard reference refers to this as "tool *name*" rather than just "tool". I think `#![register_tool_name(clippy)]` would be clear, but `#![register_attribute_tool_name(clippy)]` is quite verbose.

Author: That's not what "register" means. I think there is *some* amount of confusion possible, but I don't think it's likely enough that we need to make all the names longer. If people are adding these attributes, it's because the tool's docs told them to, and the docs should also say how to run the tool. If people are reading these attributes in existing code, then, well, I don't think it's that much more confusing than any other attribute in the language? `#[inline]` especially does not mean what most people think it means.

There is no way to disambiguate the tool; developers must use `extern crate rustfmt as my_rustfmt` (or the equivalent in Cargo.toml) to avoid the ambiguity.
As a quality of implementation issue, errors should mention one of the above workarounds.

[inert]: https://rustc-dev-guide.rust-lang.org/attributes.html#builtininert-attributes

Modules in the first path of an attribute (e.g. `#[unregistered::name]`) are assumed to be a crate if they cannot be resolved, and therefore give a hard error if not registered.

Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).

`#![no_implicit_prelude]` does not affect tools.

## Cargo

Cargo removes the current "specifying unrecognized tools may break" warning for `lints.some_unknown_tool`. Identifying unknown tools is now rustc's job, not Cargo's.

A new [manifest] section called `tools` is added. `tools` takes a list of objects in the following format:
```
NAME = { lints = BOOL, attributes = BOOL }
```
For each `NAME`:
- If `lints` is `true`, Cargo passes [`--crate-attr`]`=register_lint_tool(NAME)`

> - If `attributes` is `true`, Cargo passes `--crate-attr=register_lint_tool(NAME)`

Reviewer 9: ```suggestion
- If `attributes` is `true`, Cargo passes `--crate-attr=register_attribute_tool(NAME)`
```

# Future possibilities
[future-possibilities]: #future-possibilities

- Proc macros wish to register custom lints; see [`proc_macro_lint`]. We would have to establish some mechanism to prevent overlapping namespaces. Perhaps `warn(::project::lint_name)` could refer to the proc macro and `warn(project::lint_name)` would refer to any registered tool (only when a `project` tool is regisetered; in the common case where no tool is registered, `project::` would still refer to the proc macro).
- Projects may wish to have both a proc-macro crate with lints and a CLI with lints. To allow this, we would require `proc_macro_lint` to create an exhaustive list of lints that can be created, such that we can still run `unknown_lints` and do not need to create a new cooperation mechanism between `proc_macro_lint` and `register_lint_tool`, nor to require users of the project to distinguish the two with `::project` (see immediately above). We might still run into difficulty if the proc-macro lint namespace is only active while the proc-macro is expanding; it depends on how `proc_macro_lint` is specified. But I think it's ok to delay that discussion until `proc_macro_lint` gets an RFC.

> - We could allow proc-macros to register a scoped tool, such that e.g. `#[serde::flatten]` is valid while the proc-macro is expanding, but not elsewhere in the crate. This is similar to [derive helpers], but namespaced. We would have to take care to avoid ambiguity between the scoped tool and globally registered tools in such a way that external tools still do not need to perform name resolution.

Author: @ssbr i'm going to respond to you here so we can have things in threads; i think derive helpers are basically what you're asking for.

> Now, with the proposal, AIUI, to have a similar workflow using tool attributes, the user must both depend on the attribute macro crate, as well as independently register the tool attribute which the proc macro will generate. This is unfortunate -- the user needs to do two things, and needs to know the name of the tool attribute even if they never spell it.

so, this reminds me strongly of [derive macro helpers](https://doc.rust-lang.org/nightly/reference/procedural-macros.html#derive-macro-helper-attributes). i double checked just now and derive helpers *are* [preserved by macro-expansion](https://github.com/user-attachments/assets/136f9052-5017-4495-b7b5-e61c2e420f8b). so i expect you to continue to be able to use your proc-macro after this is stabilized, it would just use a different (non-namespaced) syntax. that seems ok - extending derive helpers to a namespace is a nice addition, but i don't think it's strictly necessary for you to keep using this. i don't see a need for proc macros to use the same syntax as hand-written code.

Reviewer 10: Are derive macro helpers also able to be emitted by, well, non-derive macros? They should work, I don't think we tried them.

>  i don't see a need for proc macros to use the same syntax as hand-written code.

+1

Author: i'm not sure if they are today, but that seems like a very reasonable extension RFC. i am not going to write that RFC but you are welcome to and i can give you pointers on how to get started.

i'll add this to future extensions.

To disambiguate a crate, developers may write `#[::rustfmt::skip]`, which forces resolution in the [Extern prelude].
To disambiguate an item in the standard prelude (e.g. `Clone`), developers may use `std::prelude::v1::Clone`.
There is no way to disambiguate the tool; developers must use `extern crate rustfmt as my_rustfmt` (or the equivalent in Cargo.toml) to avoid the ambiguity.
As a quality of implementation issue, errors should mention one of the above workarounds.

[inert]: https://rustc-dev-guide.rust-lang.org/attributes.html#builtininert-attributes

Modules in the first path of an attribute (e.g. `#[unregistered::name]`) are assumed to be a crate if they cannot be resolved, and therefore give a hard error if not registered.

Unknown tool names in lints remain a hard error until the story for proc-macro lints is resolved (see [Future possibilities](#future-possibilities)).

[Inner attributes][custom-inner-attributes] for tools remain a hard error.
Tool names are not affected by the current scope,
but we need to know the current scope in order to give an ambiguity error.

`#![no_implicit_prelude]` does not affect tools,

> i.e. the tool prelude is always present.

Reviewer 11: ```suggestion
Today, `#![no_implicit_prelude]` suppresses the tool attribute prelude but does not suppress the lint prelude (e.g., `#[allow(clippy::foo)]` continues to work). Under this RFC, neither prelude is suppressed by `#![no_implicit_prelude]`.  This is a breaking change since it can produce new ambiguity errors in stable code.
```

To confirm, the above is what we mean?  I.e., we're changing these rules in the Reference?

> r[attributes.tool.prelude]
> Tool attributes are not available if the [`no_implicit_prelude`] attribute is used.

> r[names.preludes.no_implicit_prelude.excluded-preludes]
> The `no_implicit_prelude` attribute prevents the [standard library prelude], [extern prelude], [`macro_use` prelude], and the [tool prelude] from being brought into scope for the module and its descendants.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3808/3a355803f1bd567cc669658130e8152555d0b9d1..5684d89e9c89ec9f20e2f2b271a379d98f16d506)*

Author: That’s correct, yes. The motivation is that tools (e.g. rustfmt) already ignore this distinction in practice, so this brings the language into consistency with the implementation.

(Fixing this in rustfmt would be … challenging, since it doesn’t do name resolution.)
